File: exercises/RB101_small_problems/medium_2/tri-angles.rb
=begin
# --------------------------- Problem
  Restate the problem:

  Write a method that takes 3 angles of a triangle as arguments,
  and returns a symbol :right, :acute, :obtuse, or :invalid depending
  on whether the triangle is right, acute, obtuse, or invalid triangle.

  Input: 3 integers
  Output: symbol (:right, :acute, :obtuse, or :invalid)

# --------------------------- Test Cases + Rules

  Explicit:
    - Right: One angle of the triangle is a right angle (90 degrees)
    - Acute: All 3 angles of the triangle are less than 90 degrees
    - Obtuse: One angle is greater than 90 degrees

    - To be a valid triangle:
      - sum of angles must be 180 degrees
      - all angles must be greater than 0
      - ^ both conditions need to be met

  Implicit:
    - Assume integer valued at angles
    - & arguments specified at degrees

# --------------------------- Data Structure

# --------------------------- Scrapbook

# --------------------------- Algorithm
  /* given 3 integers */
    - return :invalid unless triangle is valid
    - return :right if the highest value is 90 degrees
    - return :acute if all angles are less than 90 degrees
    - return :obtuse

=end

def valid_triangle?(angles)
  angles.all? { |angle| angle > 0 } && angles.sum == 180
end

def triangle(a, b, c)
  angles = [a, b, c]
  return :invalid unless valid_triangle?(angles)
  return :right if angles.sort[-1] == 90
  return :acute if angles.all? { |angle| angle < 90 }
  :obtuse
end

p triangle(60, 70, 50) == :acute
p triangle(30, 90, 60) == :right
p triangle(120, 50, 10) == :obtuse
p triangle(0, 90, 90) == :invalid
p triangle(50, 50, 50) == :invalid

File: exercises/JS210_Small_Problems/Medium_2/triAngles.js
/*
---------- PROBLEM ----------

Right: One angle is a right angle (exactly 90 degrees).
Acute: All three angles are less than 90 degrees.
Obtuse: One angle is greater than 90 degrees.

Valid triangle:
  - sum of angles must be 180 degrees
  - angles must be greater than 0 degrees

Write a function that takes the three angles of a triangle as arguments and 
returns one of the following four strings representing the triangle's 
classification: 'right', 'acute', 'obtuse', or 'invalid'.

INPUT:
  - 
OUTPUT:

EXAMPLES:

----------- RULES -----------
  - Assume all 3 args will be provided
  - Assume all args are integer values
  - All angles must sum to 180 AND they all must be greater than 0
  - If one angle is 90 degrees, return 'right'
  - If all three angles are less than 90 degrees, return 'acute'
  - If one angle is greater than 90 degrees, return 'obtuse'

------- DATA STRUCTURE ------
  - Use arrays for iteration methods
    - reduce
    - some()
    - every()
    - filter()

--------- SCRATCH PAD -------

---------- ALGORITHM --------
  - If sum of all angles are not 180º and any value is 0
    - return invalid
  - if one angle is 90
    - return right
  - if all angles are less than 90º
    - return acute
  - otherwise return 'obtuse'

*/

function triangle(...angles) {
  if (invalidTriangle(angles)) return 'invalid';
  
  if (angles.some(angle => angle === 90)) return 'right';
  if (angles.every(angle => angle <= 90)) return 'acute';
  return 'obtuse';
}

function invalidTriangle(angles) {
  const total = angles.reduce((total, angle) => angle + total, 0);
  const zeroValues = angles.filter(angle => angle === 0);
  return (total !== 180 || zeroValues.length > 0);
}

console.log(triangle(60, 70, 50));       // "acute"
console.log(triangle(30, 90, 60));       // "right"
console.log(triangle(120, 50, 10));      // "obtuse"
console.log(triangle(0, 90, 90));        // "invalid"
console.log(triangle(50, 50, 50));       // "invalid"